chained/water_version_control.go

```diff
@@ -0,0 +1,274 @@
+package chained
+
+import (
+	"context"
+	"io"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/getlantern/flashlight/v7/proxied"
+	"github.com/gocarina/gocsv"
+)
+
+type versionControl struct {
+	// contain all wasm files available locally
+	// key is something like: protocol.version
+	wasmFilesAvailable map[string]wasmInfo
+	dir                string
+
+	// history of the last time the WASM file was loaded
+	history      []history
```

Comment:
You're already maintaining a map of transport names to info about the transport.  Why not just track this information there?  This would simplify the logic quite a bit.  Combining this with a simplification of `wasmInfo` (suggested below), you'd have:

```
type versionControl struct {
    dir string
    wasms map[string]wasmInfo
}

type wasmInfo struct {
    lastLoaded time.Time
    path string
}
```

If you'd like, you could simplify it even further by just relying on the directory structure.  If your directory looked like:
```
dir
    ├── transport-foo.wasm
    ├── transport-foo.last-loaded
    ├── transport-bar.wasm
    └── transport-bar.last-loaded
```

Then your version control system could just be:
```
type versionControl string

// Essentially just documenting that versionControl represents the directory.
func newVersionControl(dir string) versionControl {
    return versionControl(dir)
}

func (vc versionControl) getWASM(ctx context.Context, transport string, downloader wasmDownloader) (io.ReadCloser, error) {
    path := filepath.Join(vc, transport + ".wasm")
    f, err := os.Open(path)
    if errors.Is(err, fs.ErrNotExist) {
        f, err = os.Create(path)
        // check and handle err
        err = downloader.downloadWASM(ctx, f)
        // check and handle err
        _ , err = f.Seek(0, 0)
        // check and handle err
    }
    vc.markUsed(transport)
    return f, nil
}

func (vc versionControl) markUsed(transport string) {
    f, err := os.Create(filepath.Join(vc, transport + ".last-loaded"))
    // check and log err
    // write time.Now() to f
    go vc.cleanOutdated()
}

func (vc versionControl) cleanOutdated() {
    // Walk the directory, parsing ".last-loaded" files and deleting ".wasm" files as necessary.
}
```